mythril_core/src/vga_logger.rs

```diff
@@ -0,0 +1,101 @@
+use core::fmt;
+
+const VGA_BASE_ADDR: usize = 0xB8000;
+const VGA_WIDTH: usize = 80;
+const VGA_HEIGHT: usize = 25;
+const VGA_ATTRIB: u16 = 0x0F00; // black background, white text
+
+fn get_vga_addr(x: usize, y: usize) -> usize {
+    VGA_BASE_ADDR + 2 * (y * VGA_WIDTH + x)
+}
+
+unsafe fn scroll_vga() {
+    for y in 1..VGA_HEIGHT {
+        for x in 0..VGA_WIDTH {
+            let write_addr = get_vga_addr(x, y - 1) as *mut u16;
+            let read_addr = get_vga_addr(x, y) as *const u16;
+
+            *write_addr = *read_addr;
+        }
+    }
+    clear_line_vga(VGA_HEIGHT - 1);
+}
+
+unsafe fn clear_line_vga(y: usize) {
+    for x in 0..VGA_WIDTH {
+        let addr = get_vga_addr(x, y) as *mut u16;
+        *addr = VGA_ATTRIB | 0x20;
+    }
+}
+
+pub unsafe fn clear_vga() {
+    for y in 0..VGA_HEIGHT {
+        clear_line_vga(y);
+    }
+}
+
+pub unsafe fn raw_write_vga(s: &str, x: usize, y: usize) -> (usize, usize) {
+    let mut x = x;
+    let mut y = y;
+
+    for byte in s.bytes() {
+        // move cursor on newlines (0x0A) and carriage-returns (0x0D)
+        if byte == 0x0A {
+            y += 1;
+            x = 0;
+            continue;
+        } else if byte == 0x0D {
+            x = 0;
+            continue;
+        }
+
+        if y >= VGA_HEIGHT {
+            scroll_vga();
+            y = VGA_HEIGHT - 1;
+        }
+
+        let addr = get_vga_addr(x, y) as *mut u16;
+        *addr = VGA_ATTRIB | (byte as u16);
+
+        x += 1;
+
+        if x >= VGA_WIDTH {
+            y += 1;
+            x = 0;
+        }
+    }
+
+    (x, y)
+}
+
+pub struct VgaWriter {
+    cur_x: usize,
+    cur_y: usize,
+}
+
+impl VgaWriter {
+    pub const fn new() -> Self {
+        VgaWriter { cur_x: 0, cur_y: 0 }
+    }
+
+    pub unsafe fn write(&mut self, s: &str) {
+        if self.cur_x == 0 && self.cur_y == 0 {
+            clear_vga();
+        }
+        let (x, y) = raw_write_vga(s, self.cur_x, self.cur_y);
+        self.cur_x = x;
+        self.cur_y = y;
+    }
+}
+
+impl fmt::Write for VgaWriter {
+    fn write_str(&mut self, s: &str) -> fmt::Result {
+        unsafe { self.write(s) };
+        Ok(())
+    }
+
+    fn write_fmt(&mut self, args: fmt::Arguments) -> Result<(), fmt::Error> {
+        let ret = fmt::write(self, args);
```

Comment:
This can just be `fmt::write(self, args)` (rather than storing then returning `ret`)